bee-signing/src/ternary/wots/mod.rs

```diff
@@ -0,0 +1,307 @@
+    /// Low security.
+    Low = 1,
+    /// Medium security.
+    Medium = 2,
+    /// High security.
+    High = 3,
+}
+
+impl Default for WotsSecurityLevel {
+    fn default() -> Self {
+        WotsSecurityLevel::Medium
+    }
+}
+
+/// Winternitz One Time Signature private key.
+#[derive(SecretDebug, SecretDisplay, SecretDrop)]
+pub struct WotsPrivateKey<S> {
+    pub(crate) state: TritBuf<T1B1Buf>,
+    pub(crate) sponge: PhantomData<S>,
+}
+
+impl<S> Zeroize for WotsPrivateKey<S> {
+    fn zeroize(&mut self) {
+        // This unsafe is fine since we only reset the whole buffer with zeros, there is no alignement issues.
+        unsafe { self.state.as_i8_slice_mut().zeroize() }
+    }
+}
+
+impl<S: Sponge + Default> PrivateKey for WotsPrivateKey<S> {
+    type PublicKey = WotsPublicKey<S>;
+    type Signature = WotsSignature<S>;
+    type Error = Error;
+
+    fn generate_public_key(&self) -> Result<Self::PublicKey, Self::Error> {
+        let mut sponge = S::default();
+        let mut hashed_private_key = self.state.clone();
+        let security = self.state.len() / SIGNATURE_FRAGMENT_LENGTH;
+        let mut digests = TritBuf::<T1B1Buf>::zeros(security * HASH_LENGTH);
+        let mut public_key_state = TritBuf::<T1B1Buf>::zeros(HASH_LENGTH);
+
+        // Hash each chunk of the private key the maximum amount of times.
+        for chunk in hashed_private_key.chunks_mut(HASH_LENGTH) {
+            for _ in 0..Tryte::MAX_VALUE as i8 - Tryte::MIN_VALUE as i8 {
+                sponge.absorb(chunk).map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge
+                    .squeeze_into(chunk)
+                    .map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge.reset();
+            }
+        }
+
+        // Create one digest per fragment of the private key.
+        for (i, chunk) in hashed_private_key.chunks(SIGNATURE_FRAGMENT_LENGTH).enumerate() {
+            sponge
+                .digest_into(chunk, &mut digests[i * HASH_LENGTH..(i + 1) * HASH_LENGTH])
+                .map_err(|_| Self::Error::FailedSpongeOperation)?;
+        }
+
+        // Hash the digests together to produce the public key.
+        sponge
+            .digest_into(&digests, &mut public_key_state)
+            .map_err(|_| Self::Error::FailedSpongeOperation)?;
+
+        Ok(Self::PublicKey {
+            state: public_key_state,
+            sponge: PhantomData,
+        })
+    }
+
+    fn sign(&mut self, message: &Trits<T1B1>) -> Result<Self::Signature, Self::Error> {
+        if message.len() != HASH_LENGTH {
+            return Err(Error::InvalidMessageLength(message.len()));
+        }
+
+        let mut sponge = S::default();
+        let mut signature = self.state.clone();
+
+        for (i, chunk) in signature.chunks_mut(HASH_LENGTH).enumerate() {
+            // Safe to unwrap because 3 trits can't underflow/overflow an i8.
+            let val = i8::try_from(&message[i * 3..i * 3 + 3]).unwrap();
```

Comment:
You can cast `Btrit` to `i8` using `x as i8`

Reply:
But this is not a `Btrit`, this is a `&Trits[_]`